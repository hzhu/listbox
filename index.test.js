import React from "react";
import {
  render,
  prettyDOM,
  fireEvent,
  cleanup,
  wait
} from "react-testing-library";
import "jest-dom/extend-expect";
import { Listbox, Option } from "./index";

afterEach(cleanup);

test("should render <Listbox> that has the first <Option> selected", () => {
  // Given
  const { getByLabelText, getByText, getByTestId, container } = render(
    <Listbox>
      {["Apple", "Bananna", "Carrot"].map(element => (
        <Option key={element}>
          <div
            style={{
              color: "black",
              padding: "0 1em"
            }}
          >
            {element}
          </div>
        </Option>
      ))}
    </Listbox>
  );
  const node = getByTestId("Listbox-ul");
  const eventProperties = {
    key: "ArrowDown",
    keyCode: 40,
    which: 40
  };

  // When
  fireEvent.keyDown(node, eventProperties);

  // Then
  expect(node).toMatchSnapshot();
});

test("is focused and selects first <Option> on mount when passed the focus prop", () => {
  const { getByTestId } = render(
    <Listbox focused>
      {["Apple", "Bananna", "Carrot"].map(element => (
        <Option key={element}>{element}</Option>
      ))}
    </Listbox>
  );
  const testId = "Listbox-ul";
  const node = getByTestId(testId);
  expect(node).toHaveAttribute("aria-activedescendant", "listbox-option-0");
  expect(document.activeElement.dataset.testid).toBe(testId);
});

test("should select the second option using arrow key navigation", () => {
  // Given
  const CARROT = "Carrot";
  const fruits = ["Apple", "Bananna", CARROT];
  const { getByText, getByTestId } = render(
    <Listbox>
      {fruits.map(element => (
        <Option key={element}>{element}</Option>
      ))}
    </Listbox>
  );
  const node = getByTestId("Listbox-ul");
  const eventProperties = {
    key: "ArrowDown",
    keyCode: 40,
    which: 40
  };

  // When
  fireEvent.keyDown(node, eventProperties);
  fireEvent.keyDown(node, eventProperties);
  fireEvent.keyDown(node, eventProperties);

  // Then
  expect(node).toHaveAttribute("aria-activedescendant", "listbox-option-2");
  expect(getByText(CARROT)).toHaveAttribute("aria-selected", "true");
});

test("selects and focuses the next option that starts with the typed character", () => {
  // Given
  const BANANNA = "Bananna";
  const fruits = ["Apple", BANANNA, "Carrot"];
  const { getByTestId } = render(
    <Listbox>
      {fruits.map(fruit => (
        <Option key={fruit}>{fruit}</Option>
      ))}
    </Listbox>
  );
  const node = getByTestId("Listbox-ul");
  const eventProperties = {
    key: "b",
    which: 66,
    keyCode: 66
  };

  // When
  fireEvent.keyDown(node, eventProperties);

  // Then
  const activeIdx = fruits.indexOf(BANANNA);
  expect(getByTestId("Listbox-ul").getAttribute("aria-activedescendant")).toBe(
    `listbox-option-${activeIdx}`
  );
});
