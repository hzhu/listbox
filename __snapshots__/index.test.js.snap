// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render listbox with the first option selected (snapshot) 1`] = `
<div
  aria-activedescendant="listbox__option__0-1"
  data-testid="Listbox"
  role="listbox"
  tabindex="0"
>
  <ul
    style="padding: 0px; margin: 0px;"
  >
    <li
      data-index="0"
      id="listbox__option__0-0"
      role="option"
      style="list-style: none;"
    >
      <div
        style="color: black;"
      >
        Apple
      </div>
    </li>
    <li
      aria-selected="true"
      data-index="1"
      id="listbox__option__0-1"
      role="option"
      style="list-style: none; background: rgb(189, 228, 255);"
    >
      <div
        style="color: black;"
      >
        Bananna
      </div>
    </li>
    <li
      data-index="2"
      id="listbox__option__0-2"
      role="option"
      style="list-style: none;"
    >
      <div
        style="color: black;"
      >
        Carrot
      </div>
    </li>
  </ul>
</div>
`;
